### PR TITLE
Adds doxygen modules to OpenCL and write more documentation

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This tidies up the OpenCL docs a bit and adds moduels for the kernel executor, `matrix_cl`, and error checking functions.

There's also a couple teeny fixes. Like in `opencl_context_base` we now have devices be a vector since the accessor in `opencl_context` would just create a vector anyway.

## Tests

Run `make doxygen` to check out the new OpenCL modules

## Side Effects

Prettier docs!

## Release notes

Adds documentation for the OpenCL modules.

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Thanks! Going to take a look tomorrow.

**Comment by rok-cesnovar** [comment]
@bob-carpenter @bbbales2 This PR only touches docs but an upstream (Stan repo) error occured on the Unit tests.

The tests ran on a Linux AWS instance. See logs: https://jenkins.mc-stan.org/blue/rest/organizations/jenkins/pipelines/Stan/branches/downstream_tests/runs/1432/nodes/95/steps/107/log/?start=0

**Comment by SteveBronder** [comment]
It looks like complex is doing something since it supports implicit conversion?

**Comment by rok-cesnovar** [comment]
Will restart the tests an hope it chooses the other machine.

**Comment by SteveBronder** [comment]
@serban-nicusor-toptal @rok-cesnovar what is the setup for that machine? We should probably check that this isn't a goof

**Comment by bbbales2** [comment]
@rok-cesnovar this looks like a similar error as we were getting for mingw.

In that pull request, the compiler cmath was not defining:

```
double pow(double x, int y)
```

And so we had to include an ifdef'd version: https://github.com/stan-dev/math/pull/1774/commits/1a64c9a59dcdf1bde8237ca29a9d88aba44962f1

I'm running Ubuntu 18.04 and have clang 6.0 on my computer though, and I can clone a copy of the repo and run:

```
make test/unit/optimization/bfgs_minimizer_test
```

And it works fine.

If I look in ```/usr/include/x86_64-linux-gnu/bits/mathcalls.h``` on my computer I only see a double/double version of pow, so I'm not sure where the double/int version is coming from.

Yeah I think this is a problem we should figure out. Is there an image for this EC2 instance I can use to see what's up?

**Comment by rok-cesnovar** [comment]
> what is the setup for that machine?

Its an AWS EC2 instance running Linux.

> Yeah I think this is a problem we should figure out. Is there an image for this EC2 instance I can use to see what's up?

@serban-nicusor-toptal can hook you up. That would probably be the best yeah.

**Comment by serban-nicusor-toptal** [comment]
Hey @bbbales2 please send me an email to `serban.nicusor@toptal.com` and I'll spin up a machine so you can test it out.

File: stan/math/opencl/opencl.hpp [hunk 1]
```diff
@@ -4,11 +4,66 @@
...
+ * Eigen::MatrixXd A_eig = Eigen::MatrixXd::Random(10, 10);
+ * matrix_cl<double> A(A_eig);
+ * matrix_cl<double> B = to_matrix_cl(A_eig);
+ * matrix_cl<double> C = cholesky_decompose(A * B);
+ * // Read back to eigen matrix.
+ * Eigen::MatrixXd C_eig = from_matrix_cl(C);
+ *```
+ *
+ * Execution is performed in async and Kernel operations are compounded and
+ *compiled Just In Time. This allows for a low amount of overhead when passing
```
Comment (rok-cesnovar) [inline stan/math/opencl/opencl.hpp]: We want spaces between the text and * right? Please fix those.

**Comment by bbbales2** [comment]
I think this pow thing is a real issue. I'm gonna open up something separate for it. Thanks for finding this!

**Comment by bob-carpenter** [comment]
It's that crazy Rosenbrock function example again.  But that's just getting code generated from the Stan transpiler, right?  It's not built in somehow.

File: stan/math/opencl/opencl_context.hpp [hunk 2]
```diff
@@ -139,10 +140,10 @@ class opencl_context_base {
   cl::Context context_;  // Manages the the device, queue, platform, memory, etc
   cl::CommandQueue command_queue_;       // job queue for device, one per device
   std::vector<cl::Platform> platforms_;  // Vector of available platforms
-  cl::Platform platform_;                // The platform for compiling kernels
+  std::vector<cl::Platform> platform_;   // The platform for compiling kernels
   std::string platform_name_;  // The platform such as NVIDIA OpenCL or AMD SDK
+  std::vector<cl::Device> device_;   // The selected OpenCL device
```
Comment (t4c1) [inline stan/math/opencl/opencl_context.hpp]: Why are these two now vectors?

Comment (SteveBronder) [reply stan/math/opencl/opencl_context.hpp:145]: This looks odd but if you look at the accessor below we actually treat this like a vector already, but we'd essentially make a vector each time we call `device()`

https://github.com/stan-dev/math/pull/1804/files#diff-b87cced6b3cfba6bae2714b78737a40cR379

This whole thing was essentially the first thing I ever wrote in Stan lol

**Comment by SteveBronder** [comment]
Yeah let's hold off

**Comment by SteveBronder** [comment]
@rok-cesnovar are we cool to start merging stuff?

**Comment by rok-cesnovar** [comment]
We can start merging right @serban-nicusor-toptal ? You have the tarballs for cmdstan ready just waiting for stanc3.

**Comment by serban-nicusor-toptal** [comment]
Hey, the tag for both cmdstan and stanc3 is there tied to the today last commit id so I assume we're good to go.

**Pull request merged**